atlassian_jwt_auth/signer.py

```diff
@@ -54,15 +55,73 @@ def generate_jwt(self, audience, **kwargs):
             headers={'kid': key_identifier.key_id})
 
 
+class TokenReusingJWTAuthSigner(JWTAuthSigner):
+
+    def __init__(self, issuer, private_key_retriever, **kwargs):
+        super(TokenReusingJWTAuthSigner, self).__init__(
+            issuer, private_key_retriever, **kwargs)
+        self.reuse_threshold = kwargs.get('reuse_jwt_threshold', 0.95)
+
+    def get_cached_token(self, audience, **kwargs):
+        """ returns the cached token. If there is no matching cached token
+            then None is returned.
+        """
+        return getattr(self, '_previous_token', None)
+
+    def set_cached_token(self, value):
+        """ sets the cached token."""
+        self._previous_token = value
+
+    def can_reuse_token(self, existing_token, claims):
+        """ returns True if the provided existing token can be reused
+            for the claims provided.
+        """
+        if existing_token is None:
+            return False
+        existing_claims = jwt.decode(existing_token, verify=False)
+        existing_lifetime = (int(existing_claims['exp']) -
+                             int(existing_claims['iat']))
+        this_lifetime = (claims['exp'] - claims['iat']).total_seconds()
+        if existing_lifetime != this_lifetime:
+            return False
+        about_to_expire = int(existing_claims['iat']) + (
+            self.reuse_threshold * existing_lifetime)
+        if calendar.timegm(self._now().utctimetuple()) > about_to_expire:
+            return False
+        if set(claims.keys()) != set(existing_claims.keys()):
+            return False
+        for key, val in claims.items():
+            if key in ['exp', 'iat', 'jti', 'nbf']:
+                continue
+            if existing_claims[key] != val:
+                return False
+        return True
```

Comment:
This looks good to me. I agree it makes sense to keep the claims inside the Signer and not expose them to the Auth instance.